extensions/amp-analytics/0.1/amp-analytics.js

```diff
@@ -721,13 +730,55 @@ export class AmpAnalytics extends AMP.BaseElement {
   expandAndSendRequest_(request, trigger, event) {
     this.config_['vars']['requestCount']++;
     const expansionOptions = this.expansionOptions_(event, trigger);
+    /** @type {!Object<string, (DynamicVariableBindingDef|HtmlAttrBindingDef)>} */
     const dynamicBindings =
         this.getDynamicVariableBindings_(trigger, expansionOptions);
+    dynamicBindings['HTML_ATTR'] = this.htmlAttrBinding_.bind(this);
     request.send(
         this.config_['extraUrlParams'], trigger, expansionOptions,
         dynamicBindings);
   }
 
+  /**
+   * Provides a binding for getting attributes from the DOM.
+   * Most such bindings are provided in src/service/url-replacements-impl, but
+   * this one needs access to this.win.document, which if the amp-analytics
+   * tag is contained within an amp-ad tag will NOT be the parent/publisher
+   * page. Hence the need to put it here.
+   * @param {string} cssSelector Elements matching this selector will be
+   *     included, provided they have at least one of the attributeNames
+   *     set, up to a max of 10.
+   * @param attributeNames The attributes whose values will be returned.
+   * @returns {string}
+   */
+  htmlAttrBinding_(cssSelector, ...attributeNames) {
+    const HTML_ATTR_MAX_RETURN_SIZE = 10;
+    const result = [];
+    if (!cssSelector || !attributeNames || !attributeNames.length) {
+      return JSON.stringify(result);
+    }
+    try {
+      const elements = this.win.document.querySelectorAll(cssSelector);
+      for (let i = 0; elements && i < elements.length &&
+      result.length < HTML_ATTR_MAX_RETURN_SIZE; ++i) {
+        const currentResult = {};
+        attributeNames.forEach(attributeName => {
+          const attributeValue = elements[i].getAttribute(attributeName);
+          if (attributeValue) {
+            currentResult[attributeName] = attributeValue;
+          }
+        });
+        if (Object.keys(currentResult).length) {
```

Comment:
Use a boolean variable instead of heavy GC methods.

Reply:
Fixed